Service/TemplateService.php

```diff
@@ -22,79 +29,112 @@ public function replaceKeys($text, array $entities = array()): string
+        if (!$hasMatch) {
+            return $text;
+        }
+        $hasMatches = true;
 
-                    // Get all functions that should be executed
-                    $functions = explode('.', $templateReplacementValue);
+        // The value can be called recursive
+        $passingValue = $entity;
 
-                    // Execute these functions on the defined entity with the discovered arguments
-                    foreach ($functions as $function) {
-                        preg_match('/^([a-zA-Z]*)(\((.*)\))?/', $function, $result);
+        // Get a key value pair of all arguments
+        foreach ($replacementArguments as $key => $value) {
+            $replacementArguments[$value] = $matches[$key + 1];
+            $replacePattern = str_replace($replacementArguments[$key], $matches[$key + 1 ], $replacePattern);
+        }
 
-                        $functionArguments = [];
-                        $methodName = $result[1];
-                        // Get the arguments and replace them by the real values if they are present
-                        if (isset($result[3]) && $result[3] != '') {
-                            $functionArguments = explode(',', $result[3]);
-                            foreach ($functionArguments as $key => $value) {
-                                $functionArguments[$key] = $replacementArguments[$value];
-                            }
-                        }
+        // Get all functions that should be executed
+        $functions = explode('.', $templateReplacementValueCallback);
 
-                        $passingValue = call_user_func_array(array($passingValue, $methodName), $functionArguments);
-                    }
+        // Execute these functions on the defined entity with the discovered arguments
+        foreach ($functions as $function) {
+            preg_match('/^([a-zA-Z]*)(\((.*)\))?/', $function, $result);
 
-                    // Replace the pattern with the found value
-                    $text = preg_replace('/' . $replacePattern . '/', $passingValue, $text);
+            $functionArguments = [];
+            $methodName = $result[1];
+            // Get the arguments and replace them by the real values if they are present
+            if (isset($result[3]) && $result[3] != '') {
+                $functionArguments = explode(',', $result[3]);
+                foreach ($functionArguments as $key => $value) {
+                    $functionArguments[$key] = $replacementArguments[$value];
                 }
             }
-        }
 
-        // Recursivly go trough this function until no matches are found
-        if ($hasMatches) {
-            $text = $this->replaceKeys($text, $entities);
+            $passingValue = call_user_func_array(array($passingValue, $methodName), $functionArguments);
         }
 
-        return $text;
+        // Replace the pattern with the found value
+        return preg_replace('/' . $replacePattern . '/', $passingValue, $text);
```

Comment:
Maybe you should consider using preg_replace callback() to prevent duplicate preg_* calls.